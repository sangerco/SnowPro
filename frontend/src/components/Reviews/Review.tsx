import React, { useState, useEffect } from "react";
import { fetchReviewDataById } from '../../redux/actions/reviewActions';
import { connect } from 'react-redux';
import { useParams } from "react-router";
import { Container, Dimmer, Loader, Header, Divider, Image, Label, Button } from 'semantic-ui-react';
import { ReviewReplyData } from "../../redux/types/reviewReplyTypes";
import ReviewReply from "./ReviewReply";
import ReviewReplyForm from "./ReviewReplyForm";

interface ReviewData {
    id: string;
    userId: string;
    username: string;
    skiAreaSlug: string;
    skiAreaName: string;
    header: string;
    body: string;
    stars: number;
    photos: string[];
    tags: string[];
    createdAt: Date;
    replyData: ReviewReplyData[]
};

interface ReviewProps {
    review: ReviewData | null;
    loading: boolean;
    error: string | null;
    fetchReviewDataById: (id: string, slug: string) => void;
};

const Review: React.FC<ReviewProps> = ({ review, loading, error, fetchReviewDataById }) => {
    const { slug, id } = useParams();

    const [ showReviewReplyForm, setShowReviewReplyForm ] = useState(false);

    useEffect(() => {
        fetchReviewDataById(id as string, slug as string);
    }, [id, slug, fetchReviewDataById]);

    const formatDate = (date: Date) => {
        const newDate = new Date(date);
        const options: Intl.DateTimeFormatOptions = { year: 'numeric', month: 'long', day: 'numeric' };
        return newDate.toLocaleDateString('en-US', options);
    }

    if(loading) {
        return (
            <div>
                <Dimmer active>
                    <Loader>Loading...</Loader>
                </Dimmer>
            </div>
        )
    }

    if(error) return `Error! ${error}`

    if(review) {
        const date = formatDate(review.createdAt)

        return (
            <div>
                <Container textAlign='left'><Header as='h2'>{review.header}</Header></Container>
                <Container textAlign='left'><Header as='h5'>{review.skiAreaName}</Header></Container>
                <Container textAlign="left"><Header as='h6'>By {review.username}</Header></Container>                
                <Container textAlign='right'><i>{date}</i></Container>   
                    
                <Container text>   
                    <p>{review.body}</p>
                    <Divider />
                    {review.tags.length > 0 ? review.tags.map(tag => 
                    <Label.Group green>
                            <Label as='a'>{tag}</Label>
                    </Label.Group>) : <Divider />}
                    <Divider />
                    {review.photos.length > 0 ? review.photos.map(photo => 
                    <Image src={photo} fluid />) : <p>No photos yet.</p>}
                </Container>


                {review.replyData.length > 0 ? review.replyData.map(reviewReply =>
                  <ReviewReply
                        key={reviewReply.id}
                        reviewReply={reviewReply} 
                    /> ) : <Divider /> }

                <Button onClick={() => setShowReviewReplyForm(true)}>Reply to this review?</Button>
                {showReviewReplyForm && <ReviewReplyForm />}
            </div>
        )

    }

    return null;
};

const mapStateToProps = (state: any) => ({
    review: state.review.data,
    loading: state.review.loading,
    error: state.review.error
});

const mapDispatchToProps = {
    fetchReviewDataById
};

export default connect(mapStateToProps, mapDispatchToProps)(Review);