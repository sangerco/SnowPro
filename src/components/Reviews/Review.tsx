import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Link, useParams } from "react-router-dom";
import {
  Button,
  Dimmer,
  Loader,
  Icon,
  Modal,
  Rating,
  Container,
  Header,
  Divider,
  Image,
  Grid,
  Segment,
  Card,
} from "semantic-ui-react";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchReview, deleteReview } from "../../redux/slices/reviewSlice";
import { fetchSkiAreas } from "../../redux/slices/skiAreaSlice";
import ReviewReplyForm from "./ReviewReplyForm";

const Review: React.FC = () => {
  const auth = useSelector((state: RootState) => state.auth);
  const authId = auth.data?.id;
  const isAuthenticated = auth.isAuthenticated;
  const { id } = useParams();
  const dispatch = useDispatch<AppDispatch>();
  const reviews = useSelector((state: RootState) => state.reviews);
  const [showDeleteModal, setShowDeleteModal] = useState(false);
  const [showReviewReplyModal, setShowReviewReplyModal] = useState(false);

  useEffect(() => {
    if (id) {
      dispatch(fetchReview(id));
    }
  }, [dispatch, id]);

  const handleShowDeleteModal = () => {
    setShowDeleteModal(true);
  };

  const handleDeleteReview = () => {
    if (id) {
      dispatch(deleteReview(id));
    }
  };

  const handleShowReviewReplyModal = () => {
    setShowReviewReplyModal(true);
  };

  useEffect(() => {
    dispatch(fetchSkiAreas());
  }, [dispatch]);

  const skiAreaState = useSelector((state: RootState) => state.skiAreas);
  const skiAreas = skiAreaState.skiAreas;

  const formatDate = (date: Date) => {
    const newDate = new Date(date);
    const options: Intl.DateTimeFormatOptions = {
      year: "numeric",
      month: "long",
      day: "numeric",
    };
    return newDate.toLocaleDateString("en-US", options);
  };

  if (reviews.error) {
    return (
      <Dimmer active>
        <div>
          <p>Error! Review cannot be retrieved! {`${reviews.error}`}</p>
        </div>
      </Dimmer>
    );
  }

  if (reviews.loading) {
    return (
      <Dimmer active>
        <Loader>Loading...</Loader>
      </Dimmer>
    );
  }

  if (Array.isArray(reviews.review)) {
    return <div>Error! Review cannot be retrieved!</div>;
  } else if (reviews.review) {
    const review = reviews.review;
    const date = formatDate(review.createdAt);
    const userId = review.userId;
    return (
      <Grid>
        <Grid.Row>
          <Grid.Column width={4}>
            {skiAreaState.loading ? (
              <Dimmer active>
                <Loader>Loading...</Loader>
              </Dimmer>
            ) : skiAreaState.error ? (
              <Dimmer active>
                <Header as="h1">
                  Error! Ski Area Data cannot be retrieved! {skiAreaState.error}
                </Header>
              </Dimmer>
            ) : skiAreas && skiAreas.length > 0 ? (
              <Segment style={{ overflow: "auto", maxHeight: "75vh" }}>
                <Card id="ski-area-card">
                  {" "}
                  {/* max height overflow auto */}
                  {skiAreas.map((sa) => (
                    <Card.Content key={sa.slug}>
                      <Link to={`/ski-areas/${sa.slug}`}>{sa.name}</Link>
                    </Card.Content>
                  ))}
                </Card>
              </Segment>
            ) : null}
          </Grid.Column>
          <Grid.Column width={8}>
            <Segment>
              <div>
                <Container fluid>
                  <Header as="h1">{review.header}</Header>
                  <Divider />
                  <Header as="h3">Review of {review.skiAreaName}</Header>
                  <Divider />
                  <Header as="h5">By {review.username}</Header>
                  <Header as="h6" textAlign="right">
                    {date}
                  </Header>
                  <Rating
                    icon="star"
                    defaultRating={review.stars}
                    maxRating={5}
                    disabled
                  />
                </Container>
                {authId && authId === userId ? (
                  <Container>
                    <Link to={`/ski-areas/reviews/${review.id}/update`}>
                      <Icon name="edit" style={{ cursor: "pointer" }} />
                    </Link>
                    <Icon
                      name="trash"
                      style={{ cursor: "pointer" }}
                      onClick={handleShowDeleteModal}
                    />
                  </Container>
                ) : null}
                <Container text>
                  <p>{review.body}</p>
                  <Divider />
                  {review.photos && review.photos.length > 0 ? (
                    review.photos.map((photo) => (
                      <div>
                        <Image src={photo} fluid />
                      </div>
                    ))
                  ) : (
                    <p>No photos yet.</p>
                  )}
                </Container>
                {isAuthenticated && id ? (
                  <>
                    <Divider />
                    <Container>
                      <Button
                        color="green"
                        onClick={handleShowReviewReplyModal}>
                        Reply to this review?
                      </Button>
                    </Container>
                    <Divider />
                  </>
                ) : null}
                <Modal
                  open={showDeleteModal}
                  onClose={() => setShowDeleteModal(false)}>
                  <Modal.Content>
                    Are You Sure You Want To Delete This Review?
                  </Modal.Content>
                  <Modal.Actions>
                    <Button negative onClick={handleDeleteReview}>
                      Yes
                    </Button>
                    <Button onClick={() => setShowDeleteModal(false)}>
                      Cancel
                    </Button>
                  </Modal.Actions>
                </Modal>
                {id ? (
                  <Modal
                    open={showReviewReplyModal}
                    onClose={() => setShowReviewReplyModal(false)}>
                    <ReviewReplyForm
                      reviewId={id}
                      slug={review.skiAreaSlug}
                      username={review.username}
                      userId={review.userId}
                      closeModal={() => setShowReviewReplyModal(false)}
                    />
                  </Modal>
                ) : null}

                {review.replies && review.replies.length > 0 ? (
                  review.replies.map((reply) => (
                    <Container
                      as={Link}
                      to={`/ski-areas/reviews/replies/${reply.id}`}>
                      <Header as="h2">{reply.body}</Header>
                      <Header as="h6" textAlign="right">
                        {formatDate(reply.createdAt)}
                      </Header>
                    </Container>
                  ))
                ) : (
                  <p>No replies to this review yet.</p>
                )}
              </div>
            </Segment>
          </Grid.Column>
          <Grid.Column width={4}>
            {reviews.reviews && reviews.reviews.length > 0 ? (
              <>
                {reviews.reviews.map((review) => (
                  <Card>
                    <Card.Content key={review.id} id="review-card">
                      <Card.Header>
                        <Link to={`/ski-areas/reviews/${review.id}`}>
                          {review.header}
                        </Link>{" "}
                        <Divider />
                        {review.skiAreaName}
                      </Card.Header>
                      <Card.Description>By {review.username}</Card.Description>
                      <Card.Meta>
                        <Rating
                          icon="star"
                          defaultRating={review.stars}
                          maxRating={5}
                          disabled
                        />
                      </Card.Meta>
                      {formatDate(review.createdAt)}
                    </Card.Content>
                  </Card>
                ))}
              </>
            ) : null}
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }

  return null;
};

export default Review;
