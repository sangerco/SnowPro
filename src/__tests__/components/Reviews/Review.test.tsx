/* eslint-disable testing-library/no-node-access */
import React from "react";
import Review from "../../../components/Reviews/Review";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock from "nock";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test Review", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    message: {
      message: null,
      messages: null,
      sentMessages: null,
      loading: false,
      error: null,
    },
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review 2",
          body: "test review body 2",
          stars: 1,
          photos: [],
          createdAt: "fake created date",
          replies: [],
        },
      ],
    },
    users: {
      users: [
        {
          userId: "22",
          username: "testuser2",
        },
        {
          userId: "33",
          username: "testuser3",
        },
      ],
    },
    skiAreas: {
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Review />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Review />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("test review display", () => {
  const id = "review-1";
  const useParamsMock = jest.requireMock("react-router-dom")
    .useParams as jest.Mock;
  useParamsMock.mockReturnValue({ id });

  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    message: {
      message: null,
      messages: null,
      sentMessages: null,
      loading: false,
      error: null,
    },
    reviews: {
      review: {
        id: "review-1",
        userId: "11",
        username: "testuser",
        skiAreaSlug: "ski-area-1",
        skiAreaName: "Ski Area 1",
        header: "test review",
        body: "test review body",
        stars: 5,
        photos: [],
        createdAt: "fake created date",
        replies: [],
      },
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review 2",
          body: "test review body 2",
          stars: 1,
          photos: [],
          createdAt: "fake created date",
          replies: [],
        },
      ],
    },
    users: {
      users: [
        {
          userId: "22",
          username: "testuser2",
        },
        {
          userId: "33",
          username: "testuser3",
        },
      ],
    },
    skiAreas: {
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should display a review", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get(`/ski-areas/reviews/review-1`)
      .reply(200, {
        id: "review-1",
        userId: "11",
        username: "testuser",
        skiAreaSlug: "ski-area-1",
        skiAreaName: "Ski Area 1",
        header: "test review",
        body: "test review body",
        stars: 5,
        photos: [],
        createdAt: "fake created date",
        replies: [],
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get(`/ski-areas`)
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Review />
        </BrowserRouter>
      </Provider>
    );

    await waitFor(() => {
      const review = screen.getByTestId("review");
      expect(review).toBeInTheDocument();
    });
  });
});
