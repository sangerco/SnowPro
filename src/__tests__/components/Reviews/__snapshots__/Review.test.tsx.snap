// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Review should match snapshot 1`] = `<DocumentFragment />`;
